### PR TITLE
Settings

### PR DESCRIPTION
**Pull request by sviridchik**
done patient and guardian setting

### CODE REVIEW

**Comment by sviridchik** [comment]
Снимок экрана 2023-04-24 в 12 14 08
Снимок экрана 2023-04-24 в 12 13 50

**POST {{local}}/managment/patients/**
`{
  "first_name": "testSet",
  "last_name": "testovich",
  "age": 22,
  "phone": 375447990822
}`


**GET {{local}}/managment/settings/**
`[
    {
        "id": 2,
        "color": "white",
        "font": 14,
        "city": "Minsk",
        "language": "RUSSIAN",
        "patient": 7
    }
]`

**Comment by sviridchik** [comment]
Снимок экрана 2023-04-24 в 12 15 43
Снимок экрана 2023-04-24 в 12 18 57

**POST {{local}}/managment/guardians/**

`{
  "first_name": "guardianSet",
  "last_name": "testovich",
  "age": 22,
  "phone": 375440943822
}`

`{
    "id": 6,
    "user": {
        "id": 10,
        "email": "",
        "username": "qwe"
    },
    "first_name": "guardianSet",
    "last_name": "testovich",
    "phone": 375440943822
}`

**GET {{local}}/managment/settings_guard/**
`[
    {
        "id": 1,
        "language": "RUSSIAN",
        "theme": "white",
        "guardian": 6
    }
]`

**Comment by sviridchik** [comment]
**GET  {{local}}/statistic/report/**

Снимок экрана 2023-04-25 в 11 14 45

`{
    "taken": [
        {
            "id": 1,
            "med": {
                "id": 24,
                "title": "ярина",
                "dose": 1.0,
                "dose_type": "pcs",
                "type": "pill",
                "strict_status": false,
                "food": "Before meals",
                "patient": 5,
                "schedule": 32
            },
            "date": "2023-04-24T14:18:11.976719+03:00",
            "is_late": true,
            "report": false,
            "patient": 5
        },
        {
            "id": 2,
            "med": {
                "id": 25,
                "title": "юритмик",
                "dose": 1.0,
                "dose_type": "ml",
                "type": "ampule",
                "strict_status": false,
                "food": "Before meals",
                "patient": 5,
                "schedule": 33
            },
            "date": "2023-04-24T14:21:48.921206+03:00",
            "is_late": true,
            "report": false,
            "patient": 5
        },
        {
            "id": 3,
            "med": {
                "id": 26,
                "title": "тотема",
                "dose": 3.0,
                "dose_type": "ml",
                "type": "ampule",
                "strict_status": false,
                "food": "No matter",
                "patient": 5,
                "schedule": 34
            },
            "date": "2023-04-24T14:22:41.247434+03:00",
            "is_late": true,
            "report": false,
            "patient": 5
        },
        {
            "id": 4,
            "med": {
                "id": 27,
                "title": "missed",
                "dose": 34.0,
                "dose_type": "ml",
                "type": "pill",
                "strict_status": false,
                "food": "While eating",
                "patient": 5,
                "schedule": 35
            },
            "date": "2023-04-24T15:02:50.260921+03:00",
            "is_late": true,
            "report": false,
            "patient": 5
        },
        {
            "id": 5,
            "med": {
                "id": 27,
                "title": "missed",
                "dose": 34.0,
                "dose_type": "ml",
                "type": "pill",
                "strict_status": false,
                "food": "While eating",
                "patient": 5,
                "schedule": 35
            },
            "date": "2023-04-24T15:08:08.016498+03:00",
            "is_late": true,
            "report": false,
            "patient": 5
        },
        {
            "id": 6,
            "med": {
                "id": 27,
                "title": "missed",
                "dose": 34.0,
                "dose_type": "ml",
                "type": "pill",
                "strict_status": false,
                "food": "While eating",
                "patient": 5,
                "schedule": 35
            },
            "date": "2023-04-24T15:10:26.455849+03:00",
            "is_late": true,
            "report": false,
            "patient": 5
        }
    ],
    "missed": [
        {
            "id": 1,
            "med": {
                "id": 27,
                "title": "missed",
                "dose": 34.0,
                "dose_type": "ml",
                "type": "pill",
                "strict_status": false,
                "food": "While eating",
                "patient": 5,
                "schedule": 35
            },
            "date": "2023-04-24T15:16:37.700610+03:00",
            "is_informed": false,
            "patient": 5
        },
        {
            "id": 2,
            "med": {
                "id": 27,
                "title": "missed",
                "dose": 34.0,
                "dose_type": "ml",
                "type": "pill",
                "strict_status": false,
                "food": "While eating",
                "patient": 5,
                "schedule": 35
            },
            "date": "2023-04-24T15:16:39.447906+03:00",
            "is_informed": false,
            "patient": 5
        },
        {
            "id": 3,
            "med": {
                "id": 27,
                "title": "missed",
                "dose": 34.0,
                "dose_type": "ml",
                "type": "pill",
                "strict_status": false,
                "food": "While eating",
                "patient": 5,
                "schedule": 35
            },
            "date": "2023-04-24T15:18:21.385091+03:00",
            "is_informed": false,
            "patient": 5
        }
    ]
}`

**Comment by sviridchik** [comment]
тест с подопечным
**GET {{local}}/statistic/report/?ward=5**

Снимок экрана 2023-04-25 в 11 40 07

`{
    "taken": [
        {
            "id": 1,
            "med": {
                "id": 24,
                "title": "ярина",
                "dose": 1.0,
                "dose_type": "pcs",
                "type": "pill",
                "strict_status": false,
                "food": "Before meals",
                "patient": 5,
                "schedule": 32
            },
            "date": "2023-04-24T14:18:11.976719+03:00",
            "is_late": true,
            "report": false,
            "patient": 5
        },
        {
            "id": 2,
            "med": {
                "id": 25,
                "title": "юритмик",
                "dose": 1.0,
                "dose_type": "ml",
                "type": "ampule",
                "strict_status": false,
                "food": "Before meals",
                "patient": 5,
                "schedule": 33
            },
            "date": "2023-04-24T14:21:48.921206+03:00",
            "is_late": true,
            "report": false,
            "patient": 5
        },
        {
            "id": 3,
            "med": {
                "id": 26,
                "title": "тотема",
                "dose": 3.0,
                "dose_type": "ml",
                "type": "ampule",
                "strict_status": false,
                "food": "No matter",
                "patient": 5,
                "schedule": 34
            },
            "date": "2023-04-24T14:22:41.247434+03:00",
            "is_late": true,
            "report": false,
            "patient": 5
        },
        {
            "id": 4,
            "med": {
                "id": 27,
                "title": "missed",
                "dose": 34.0,
                "dose_type": "ml",
                "type": "pill",
                "strict_status": false,
                "food": "While eating",
                "patient": 5,
                "schedule": 35
            },
            "date": "2023-04-24T15:02:50.260921+03:00",
            "is_late": true,
            "report": false,
            "patient": 5
        },
        {
            "id": 5,
            "med": {
                "id": 27,
                "title": "missed",
                "dose": 34.0,
                "dose_type": "ml",
                "type": "pill",
                "strict_status": false,
                "food": "While eating",
                "patient": 5,
                "schedule": 35
            },
            "date": "2023-04-24T15:08:08.016498+03:00",
            "is_late": true,
            "report": false,
            "patient": 5
        },
        {
            "id": 6,
            "med": {
                "id": 27,
                "title": "missed",
                "dose": 34.0,
                "dose_type": "ml",
                "type": "pill",
                "strict_status": false,
                "food": "While eating",
                "patient": 5,
                "schedule": 35
            },
            "date": "2023-04-24T15:10:26.455849+03:00",
            "is_late": true,
            "report": false,
            "patient": 5
        }
    ],
    "missed": [
        {
            "id": 1,
            "med": {
                "id": 27,
                "title": "missed",
                "dose": 34.0,
                "dose_type": "ml",
                "type": "pill",
                "strict_status": false,
                "food": "While eating",
                "patient": 5,
                "schedule": 35
            },
            "date": "2023-04-24T15:16:37.700610+03:00",
            "is_informed": false,
            "patient": 5
        },
        {
            "id": 2,
            "med": {
                "id": 27,
                "title": "missed",
                "dose": 34.0,
                "dose_type": "ml",
                "type": "pill",
                "strict_status": false,
                "food": "While eating",
                "patient": 5,
                "schedule": 35
            },
            "date": "2023-04-24T15:16:39.447906+03:00",
            "is_informed": false,
            "patient": 5
        },
        {
            "id": 3,
            "med": {
                "id": 27,
                "title": "missed",
                "dose": 34.0,
                "dose_type": "ml",
                "type": "pill",
                "strict_status": false,
                "food": "While eating",
                "patient": 5,
                "schedule": 35
            },
            "date": "2023-04-24T15:18:21.385091+03:00",
            "is_informed": false,
            "patient": 5
        }
    ]
}`

File: api/managment/views.py [hunk 1]
```diff
@@ -78,7 +84,31 @@ class PatientSettingViewSet(viewsets.ModelViewSet):
...
+    def get_object(self):
+        queryset = self.filter_queryset(self.get_queryset())
+        return queryset.get(patient=self.request.user.patient)
 
+    def get_queryset(self):
+        user: User = self.request.user
+        if user.is_superuser:
+            return PatientSetting.objects.all()
+        else:
+            return PatientSetting.objects.filter(patient=user.patient)
```
Comment (AHTOH2001) [inline api/managment/views.py]: Should be 404 instead of 500
image

File: api/choices.py
Comment (AHTOH2001) [inline api/choices.py:1]: image

**But**
image

File: api/managment/views.py [hunk 2]
```diff
@@ -78,7 +84,31 @@ class PatientSettingViewSet(viewsets.ModelViewSet):
...
     serializer_class = PatientSettingSerializer
     permission_classes = [IsAuthenticated]
+    def get_object(self):
+        queryset = self.filter_queryset(self.get_queryset())
+        return queryset.get(patient=self.request.user.patient)
 
+    def get_queryset(self):
+        user: User = self.request.user
+        if user.is_superuser:
+            return PatientSetting.objects.all()
```
Comment (AHTOH2001) [inline api/managment/views.py]: Just notice that we don't need logic of superuser. But you can leave this code

File: api/choices.py
Comment (AHTOH2001) [reply api/choices.py:1]: image

**But**
image

**Review by AHTOH2001 — CHANGES_REQUESTED** [review]
Also:
1. Remove all code comments
2. Merge with main

File: api/managment/serializers.py [hunk 3]
```diff
@@ -154,3 +153,15 @@ def update(self, instance, validated_data):
         instance.set_password(validated_data['password_new'])
         instance.save()
         return instance
+
+class BuySerializer(serializers.Serializer):
+    code = serializers.IntegerField(min_value=100000)
+    should_send_report = serializers.BooleanField()
+    relationship = serializers.CharField(max_length=128, min_length=8)
```
Comment (AHTOH2001) [inline api/managment/serializers.py]: Remove

File: api/managment/views.py [hunk 4]
```diff
@@ -29,16 +38,20 @@ def list(self, request, *args, **kwargs):
...
+        res = {"type": None, "user": None}
+        if len(patient) != 0:
+            res["type"] = "patient"
+            res["user"] = PatientSerializer(tuple(patient)[0]).data
+        elif len(guardian) != 0:
+            res["type"] = "guardian"
+            res["user"] = GuardianSerializer(tuple(guardian)[0]).data
+        else:
+            res["type"] = "nothing"
+            res["user"] = UserSerializer(user).data
```
Comment (AHTOH2001) [inline api/managment/views.py]: Revert as in the master
image

File: api/managment/views.py [hunk 5]
```diff
@@ -197,3 +246,35 @@ def change_password_view(request):
...
+        guardian = Guardian.objects.get(user=request.user)
+        should_send_report = request.data['should_send_report']
+        relationship = request.data['relationship']
+        count_patient_relations_of_guardians = len(PatientGuardianRelation.objects.filter(guardian=guardian))
+        count_guardians_of_patient_relations = len(PatientGuardianRelation.objects.filter(patient=patient))
+        if count_guardians_of_patient_relations<3 and count_patient_relations_of_guardians<3 and len(PatientGuardianRelation.objects.filter(patient=patient,guardian=guardian)) == 0:
+            PatientGuardianRelation.objects.create(patient=patient,guardian=guardian,should_send_report=should_send_report,relationship=relationship)
+        else:
+            return Response({'detail':"too much connections"},status=status.HTTP_400_BAD_REQUEST)
+        return Response({}, status=status.HTTP_201_CREATED)
```
Comment (AHTOH2001) [inline api/managment/views.py]: Нужно делать проверку 6-ти значного кода

File: ml.py
Comment (AHTOH2001) [inline ml.py:1]: Remove this file

**Review by AHTOH2001 — CHANGES_REQUESTED** [review]
Format with black new code

File: ml.py
Comment (sviridchik) [reply ml.py:1]: noooooo

**Review by AHTOH2001 — APPROVED** [review]

**Pull request merged**